Document microgrid concepts from an SDK perspective (#722)
frequenz-floss/frequenz-sdk-python: docs/user-guide/microgrid-concepts.md

Reviewer: I would add links to the [glossary](https://github.com/frequenz-floss/frequenz-sdk-python/blob/v1.x.x/docs/user-guide/glossary.md) the first time new terms appear at least: `{{glossary("PV")}}` for example.

Author: done

Reviewer: This might probably go well as the [`frequenz.sdk.microgrid`](https://github.com/frequenz-floss/frequenz-sdk-python/blob/v1.x.x/src/frequenz/sdk/microgrid/__init__.py) module documentation, I would put it there and then include it in `docs/user-guide/microgrid-concepts.md` using:

```md
# Microgrid Concepts

::: frequenz.sdk.microgrid
    options:
        members: []
        show_bases: false
        show_root_heading: false
        show_root_toc_entry: false
        show_source: false
```

Like we did with the actors docs.

Author: will do this as the last step, so that I can still get syntax highlighting while I make changes.

Author: done

> # Microgrid concepts

Reviewer: Nitpick as we are normally using this style.
```suggestion
# Microgrid Concepts
```

Author: done

A microgrid is a local electrical grid that connects a set of
electrical components together.  They are often built around a passive
power consumer, to supplement the electricity consumed from the public
grid, with on-site power generation or storage systems.

Microgrids can also function in island-mode, without a grid
connection, or without a local power consumer, but they have to have
at least one of the two, to be meaningful.

## Frequenz SDK Microgrid Model

> The SDK aims to provide an abstract model of the microgrid that allows
> for high-level interactions with microgrid components, without having
> to deal with, or even be aware of location specific details like:

Reviewer 2: ```suggestion
The SDK aims to provide an abstract model of the microgrid, enabling
high-level interactions with microgrid components, without the need
to handle or be aware of location-specific details such as:
```

Author: ```diff
-The SDK aims to provide an abstract model of the microgrid that allows
+The SDK aims to provide an abstract model of the microgrid, enabling
```
This changes the meaning subtly, from "the model of the microgrid" doing anything to "the sdk" doing something.  So I've changed it to "... that enables"

---

```diff
-without having to deal with, or even be aware of location specific
+without the need to handle or be aware of location-specific
```
I feel like there's a loss of emphasis with this suggestion.  Changing to "... without having to worry about (or even be aware of) location-specific ..."

- where the meters are placed,
- how many batteries,
- whether there's a grid connection or a passive consumer,
- what models the inverters are, etc.
- whether components are having downtimes, because metrics and limits
  get adjusted automatically when components are having downtimes.

Users of the SDK can develop applications around this interface once
and deploy anywhere, and the SDK will take care of translating the
requests and instructions to correspond to the specific microgrid
configurations.

``` mermaid
flowchart LR

subgraph Left[Measurements only]
direction LR
  grid["Grid Connection"]
  consumer["Consumer"]
  pv["PV Arrays"]
  chp["CHP"]
end

junction(( ))

subgraph Right[Measurements and control]
direction LR
  bat["Batteries"]
  ev["EV Chargers"]
end

grid --- junction
consumer --- junction
pv --- junction
chp --- junction

junction --- bat
junction --- ev
```

## Grid

This refers to a microgrid's connection to the external Grid.  The
power flowing through this connection can be streamed through

> [`grid_power`][frequenz.sdk.timeseries.logical_meter.LogicalMeter.grid_power].

Reviewer: Completely unrelated, but in this context, now that we have a `microgrid.grid` object, I wonder if we shouldn't publish metrics there, like we do with `battery_pool`. We can do it after 1.0 (just adding also `microgrid.grid.power` and deprecating `logical_meter.grid_power`).

Author: true, that makes sense.

Reviewer: * #729

> In locations without a grid connection, this method would still be
> available, and would stream zero values.

Reviewer 2: ```suggestion
In locations without a grid connection, this method remains
available and streams zero values.
```

Author: done, also changed the corresponding sentences for other methods to use the simple present.

## Consumer

This is the main power consumer at the site of a microgrid, and often
the load the microgrid is built to support.  The power drawn by the
consumer is available through
[`consumer_power`][frequenz.sdk.timeseries.logical_meter.LogicalMeter.consumer_power]

If there is no consumer at a site, this method would stream zero values.

## Producers: PV Arrays, CHP

The total PV power production in a microgrid can be streamed through
[`pv_power`][frequenz.sdk.timeseries.logical_meter.LogicalMeter.pv_power]
, and similarly the total CHP production in a site can be streamed
through
[`chp_power`][frequenz.sdk.timeseries.logical_meter.LogicalMeter.chp_power].
And total producer power is available through
[`producer_power`][frequenz.sdk.timeseries.logical_meter.LogicalMeter.producer_power].

> And just like the other methods, if components of either of these
> types are not available in a microgrid, the corrensponding methods
> will stream zero values.

Reviewer 2: ```suggestion
Similarly to the other methods, if components of either of these
types are unavailable in a microgrid, the corresponding methods
will stream zero values.
```

Author: Rewrote with "As is the case with the other methods, ..."

[`temperature`][frequenz.sdk.timeseries.battery_pool.BatteryPool.temperature]
and available power bounds through the
[`power_status`][frequenz.sdk.timeseries.battery_pool.BatteryPool.power_status]
method.

The `battery_pool` also provides control methods
[`propose_power`][frequenz.sdk.timeseries.battery_pool.BatteryPool.propose_power]
(which accepts values in the Passive Sign Convention and supports both
charging and discharging), or through
[`propose_charge`][frequenz.sdk.timeseries.battery_pool.BatteryPool.propose_charge],
or
[`propose_discharge`][frequenz.sdk.timeseries.battery_pool.BatteryPool.propose_discharge].

## EV Chargers

The [`ev_charger_pool`][frequenz.sdk.microgrid.ev_charger_pool] offers
a
[`power`][frequenz.sdk.timeseries.ev_charger_pool.EVChargerPool.power]
method that streams the total power measured for all the EV Chargers
at a site.

It also offers a
[`component_data`][frequenz.sdk.timeseries.ev_charger_pool.EVChargerPool.component_data]
method for fetching the status of individual EV Chargers, including
state changes like when an EV is connected or disconnected, and a
[`set_bounds`][frequenz.sdk.timeseries.ev_charger_pool.EVChargerPool.set_bounds]

> method for limit the charge power of individual EV Chargers.

Reviewer 2: Maybe `for limiting` or `to limit` instead?

Author: done
